base/strings/basic.jl

```diff
@@ -298,12 +298,13 @@ julia> cmp("b", "β")
 """
 function cmp(a::AbstractString, b::AbstractString)
     a === b && return 0
-    a, b = Iterators.Stateful(a), Iterators.Stateful(b)
-    for (c::AbstractChar, d::AbstractChar) in zip(a, b)
+    (iv1, iv2) = (iterate(a), iterate(b))
+    while iv1 !== nothing && iv2 !== nothing
+        (c, d) = (first(iv1)::AbstractChar, first(iv2)::AbstractChar)
         c ≠ d && return ifelse(c < d, -1, 1)
+        (iv1, iv2) = (iterate(a, last(iv1)), iterate(b, last(iv2)))
```

Comment:
```suggestion
        (iv1, iv2) = (iterate(a, last(iv1)::AbstractChar), iterate(b, last(iv2)::AbstractChar))
```
makes it more robust against invalidations?

Reply:
The last element of iv1 and iv2 is the state, which we do not know what is

Reply:
The last element of the iv1 and iv2 tuple is the state of the iterator, which we cannot know what type is

Reply:
I guess I could instead typeassert iv1 to be Tuple{AbstractChar, Any}

Reply:
Okay, I guess that `typeassert` can be potentially heavy, so we can leave this as is until we find this piece of code can really lead to invalidations.